library/ceph_crush.py

```diff
@@ -0,0 +1,179 @@
+  with_items: "{{ groups[osd_group_name] }}"
+  when:
+    - crush_rule_config
+'''
+
+
+def fatal(message, module):
+    '''
+    Report a fatal error and exit
+    '''
+    if module:
+        module.fail_json(msg=message, rc=1)
+    else:
+        raise(Exception(message))
+
+
+def generate_cmd(cluster, subcommand, bucket, bucket_type, containerized=None):
+    '''
+    Generate command line to execute
+    '''
+    cmd = [
+        'ceph',
+        '--cluster',
+        cluster,
+        'osd',
+        'crush',
+        subcommand,
+        bucket,
+        bucket_type,
+    ]
+    if containerized:
+        for piece in reversed(containerized.split()):
+            cmd.insert(0, piece)
+    return cmd
+
+
+def sort_osd_crush_location(location, module):
+    '''
+    Sort osd_crush_location dict
+    '''
+    len(location) < 2 and fatal("You must specify at least 2 buckets.", module)
+    [item for item in location if item[0] == "host"] or fatal("You must specify a 'host' bucket.", module)
+    try:
+        crush_bucket_types = [
+            "host",
+            "chassis",
+            "rack",
+            "row",
+            "pdu",
+            "pod",
+            "room",
+            "datacenter",
+            "region",
+            "root",
+        ]
+        return sorted(location, key=lambda crush: crush_bucket_types.index(crush[0]))
+    except ValueError as error:
+        fatal("{} is not a valid CRUSH bucket, valid bucket types are {}".format(error.args[0].split()[0], crush_bucket_types), module)
+
+
+def create_and_move_buckets_list(cluster, location, containerized=None):
+    '''
+    Creates Ceph CRUSH buckets and arrange the hierarchy
+    '''
+    previous_bucket = None
+    cmd_list = []
+    for item in location:
+        bucket_type = item[0]
+        bucket_name = item[1]
+        # ceph osd crush add-bucket maroot root
+        cmd_list.append(generate_cmd(cluster, "add-bucket", bucket_name, bucket_type, containerized))
+        if previous_bucket:
+            # ceph osd crush move monrack root=maroot
+            cmd_list.append(generate_cmd(cluster, "move", previous_bucket[1], bucket_type + "=" + bucket_name, containerized))
+        previous_bucket = item
+    return cmd_list
+
+
+def exec_commands(module, cmd_list):
+    '''
+    Creates Ceph commands
+    '''
+    for cmd in cmd_list:
+        rc, out, err = module.run_command(cmd, encoding=None)
+    return rc, cmd, out, err
+
+
+def run_module():
+    module_args = dict(
+        cluster=dict(type='str', required=False, default='ceph'),
+        location=dict(type='dict', required=True),
+        containerized=dict(type='str', required=True, default=None),
+
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
+    cluster = module.params['cluster']
+    location_dict = module.params['location']
+    location = sort_osd_crush_location(tuple(location_dict.items()), module)
```

Comment:
aha, so it is here that you convert the location. If you just used `location_dict` instead, this would be easier to implement.

The call to tuple here is redundant as well, since it is what `items()` produces:

    In [10]: location
    Out[10]:
    [('host', 'monhost'),
     ('chassis', 'monchassis'),
     ('rack', 'first'),
     ('root', 'maroute'),
     ('pod', 'monpod')]
    
    In [11]: location_dict = dict(location)
    
    In [12]: location_dict
    Out[12]:
    {'chassis': 'monchassis',
     'host': 'monhost',
     'pod': 'monpod',
     'rack': 'first',
     'root': 'maroute'}
    
    In [13]: location_dict.items()
    Out[13]:
    [('host', 'monhost'),
     ('chassis', 'monchassis'),
     ('root', 'maroute'),
     ('rack', 'first'),
     ('pod', 'monpod')]
    
    In [14]: location
    Out[14]:
    [('host', 'monhost'),
     ('chassis', 'monchassis'),
     ('rack', 'first'),
     ('root', 'maroute'),
     ('pod', 'monpod')]

Reply:
fixed in the next push